presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/dereference/PushDownDereferenceThroughProject.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule.dereference;
+
+import com.google.common.collect.HashBiMap;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.ExpressionNodeInliner;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.tree.DereferenceExpression;
+
+import java.util.Map;
+
+import static com.google.common.collect.ImmutableMap.toImmutableMap;
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.iterative.rule.dereference.PushDownDereferencesUtil.validPushdownThroughProject;
+import static io.prestosql.sql.planner.plan.Patterns.project;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Transforms:
+ * <pre>
+ *  Project(msg_x := msg.x)
+ *    Project(msg := msg)
+ *      Source(msg)
+ *  </pre>
+ * to:
+ * <pre>
+ *  Project(msg_x := symbol)
+ *    Project(msg := msg, symbol := msg.x)
+ *      Source(msg)
+ * </pre>
+ */
+public class PushDownDereferenceThroughProject
```

Comment:
> Under the current implementation of PushDownDereferenceThroughProject, we can end up in the following situation:
> 
> Given:
> 
```
1. Project(x = r0.x)
  2.  Source(r0, r1)
```
> prune project source columns will rewrite it to:
> 
```
1. Project(x = r0.x)
  2.  Project(r0)
    3.    Source(r0, r1)

```

That's not what we do. We don't prune Project source columns by inserting a pruning projection below.
The non-identity projection `x = r0.x` is a pruning projection itself.
As such, it will trigger a project-off rule on Source (to drop symbol r1). The resulting plan might be:

```
1.  Project(x = r0.x)
  2.  Source(r0)
    3.  Project(r0)
```

I think it wouldn't interfere with dereference pushdown. Neither I see a chance for an optimizer loop here. Regardless of the order of rule application (dereference pushdown and column pruning), the resulting plan is the same.
We shouldn't give up dereference pushdown when we encounter an identity projection.

Reply:
Thanks @kasiafi. I forgot the detail about not doing this for projections.